src/vdc_collection.rs

```diff
@@ -0,0 +1,272 @@
+    #[error("Failed to Deserialize Value")]
+    DeserializeFailed,
+
+    /// Attempting to write the credential to storage failed.
+    #[error("Failed to Write to Storage")]
+    StoreFailed(StorageManagerError),
+
+    /// Attempting to read the credential from storage failed.
+    #[error("Failed to Read from Storage")]
+    LoadFailed(StorageManagerError),
+
+    /// Attempting to delete a credential from storage failed.
+    #[error("Failed to Delete from Storage")]
+    DeleteFailed(StorageManagerError),
+}
+
+impl VdcCollection {
+    /// Create a new credential set.
+    pub fn new(engine: Box<dyn StorageManagerInterface>) -> VdcCollection {
+        VdcCollection { storage: engine }
+    }
+
+    /// Add a credential to the set.
+    pub fn add(
+        &self,
+        id: Uuid,
+        format: CredentialFormat,
+        ctype: CredentialType,
+        payload: Vec<u8>,
+    ) -> Result<(), VdcCollectionError> {
+        let val = match serde_cbor::to_vec(&Credential::new(id, format, ctype, payload)) {
+            Ok(x) => x,
+            Err(_) => return Err(VdcCollectionError::SerializeFailed),
+        };
+
+        match self.storage.add(self.id_to_key(id), Value(val)) {
+            Ok(()) => Ok(()),
+            Err(e) => Err(VdcCollectionError::StoreFailed(e)),
+        }
+    }
+
+    /// Get a credential from the store.
+    pub fn get(&self, id: &str) -> Result<Credential, VdcCollectionError> {
+        let raw = match self.storage.get(self.str_to_key(id)) {
+            Ok(x) => x.0,
+            Err(e) => return Err(VdcCollectionError::LoadFailed(e)),
+        };
+
+        match serde_cbor::de::from_slice(&raw) {
+            Ok(x) => Ok(x),
+            Err(_) => Err(VdcCollectionError::DeserializeFailed),
+        }
+    }
+
+    /// Remove a credential from the store.
+    pub fn delete(&self, id: &str) -> Result<(), VdcCollectionError> {
+        match self.storage.remove(self.str_to_key(id)) {
+            Ok(_) => Ok(()),
+            Err(e) => Err(VdcCollectionError::DeleteFailed(e)),
+        }
+    }
+
+    /// Get a list of all the credentials.
+    pub fn all_entries(&self) -> Vec<Key> {
+        let mut r = Vec::new();
+
+        for key in self.storage.list() {
+            let name = key.0;
+
+            if name.starts_with(KEY_PREFIX) {
+                if let Some(id) = name.strip_prefix(KEY_PREFIX) {
+                    r.push(Key(id.to_string()));
+                }
+            }
+        }
+
+        r
+    }
+
+    /// Get a list of all the credentials that match a specified type.
+    pub fn all_entries_by_type(&self, ctype: CredentialType) -> Vec<String> {
+        let mut r = Vec::new();
+
+        for key in self.all_entries() {
+            let cred = self.get(&key.0);
+
+            if let Ok(x) = cred {
+                if x.ctype == ctype {
+                    r.push(key.0);
+                }
+            }
+        }
+
+        r
+    }
```

Comment:
Edit: I noticed that we're returning `key.0`, so the above example would not work.

The following example returns the inner Key String. One question maybe worth asking is whether we should return the `Key` type and have a `.deref()` method for the inner String.

```rust
self.all_entries()
            .iter()
            .filter_map(|key| {
                if let Ok(cred) = self.get(&key.0) {
                    if cred.ctype == ctype {
                        return Some(key.0.clone());
                    }
                }

                None
            })
            .collect()
```

Reply:
I'm not entirely convinced the `Key` type is doing anything for us, given it's just a wrapper around `String`. It was in the original interface, so I left it as it was, but I don't see any benefit to keeping it as opposed to simply using `String` for keys.

Reply:
It's just syntactic sugar for a type alias when compiled into Swift and Kotlin. Just makes the API a bit easier to use, instead of having to think about "what string am I supposed to use"?

> public typealias Key = String